### ISSUE TITLE
Cannot flash bootloader image.

### ISSUE THREAD

**Issue by MaciejMalczyk** [tofurky/tegra30_debrick#6]
I have problem with step 7 - nvflash downloads bootloader image  to Nexus tablet but after reboot it does not boot into bootloader even with pwr + vol down.  Instead it goes to APX mode. Device have broken system image and it will not boot automatically into system after bootloader flash. Is there any way overcome this? Or maybe i'm doing something wrong?

The only difference that i have in comparison to debrick instructions is flag: "sdram config strap: 1"  from 6th step.

**Comment by tofurky** [comment]
have you tried appending --sync to the command in step 7?

**Comment by MaciejMalczyk** [comment]
Yes. Unfortunatelly it didn't help. Nvflash hangs at "bootloader/bootloader-grouper-4.23.img sent successfully" and after force reboot to bootloader mode device falls to APX.

**Comment by tofurky** [comment]
it shouldn't hang like that. uart would be useful to see what is going on there.

is this by any chance the 3g version (tilapia / nakasig) ? or is it the wifi (grouper / nakasi)?

**Comment by MaciejMalczyk** [comment]
This is for sure grouper. When the bootloader was working normally the codename grouper was on the screen. Tablet went to APX state after my trials to unlock bootlooder. Device had btoken system partition (but for 99% logicaly, not physicaly because i can tried emmc read test from other issue here but for morze than 2GB of disk space).
Maybe i should format partition where system is written and than try to flash bootlooder? Is it even possible to do with nvflash?

I will try to find out where is and how to connect to uart on this device (propably with PostmarketOS wiki). What i should expect on uart tty?

**Comment by tofurky** [comment]
i don't think that flashing system will prevent APX mode, system should only come into play after the kernel boots.

after step 6, you could try uploading zeroes to the LNX partition like shown in step 9 of the ouya debrick. maybe it's possible that it's hanging trying to boot the existing kernel on there, UART would show this definitively though.

```
truncate -s 8M LNX_all_zeroes.bin
sudo ./utils/nvflash_v1.13.87205_miniloader_patched --resume --download LNX LNX_all_zeroes.bin --go
```

in the ouya's case, it automatically falls back to standard fastboot mode at that point. i'm unsure how the n7 will behave there.
 
if you wind up in fastboot mode (`dmesg -Tw` will show this), you could then try running:
```
sudo fastboot flash bootloader bootloader/bootloader-grouper-4.23.img
```

if it didn't go to fastboot mode automatically, try step 7 again with --sync. note you should only need to overwrite the LNX (kernel) partition with zeroes once in order to see if step 7 stops hanging.



n7 uart isn't documented anywhere afaik, i found a schematic floating around the net for the ME370T (http://www.s-manuals.com/pdf/motherboard/asus/asus_me370t_r1.0_schematics.pdf) see:

https://github.com/tofurky/tegra30_debrick/blob/master/image/n7_uart.jpg
https://github.com/tofurky/tegra30_debrick/blob/master/image/n7_uart_schematic.jpg
https://github.com/tofurky/tegra30_debrick/blob/master/image/n7_uart_context.jpg

if you do try this, be careful though, the pads on the pcb are quite fragile and any amount of tugging on the soldered wires may just rip them clear off the board.

by default there is no uart output, using this patched bootloader will give output:
https://github.com/tofurky/tegra30_debrick/blob/master/bootloader/bootloader-grouper-4.23_uart.img

you shouldn't actually write bootloader-grouper-4.23_uart.img to the EBT partition, just use it for step 6 i.e.:
```
sudo ./utils/nvflash_v1.13.87205_miniloader_patched --setbct --bct ./bct/nexus_7_grouper_bct.bin --configfile ./utils/flash.cfg --bl ./bootloader/bootloader-grouper-4.23_uart.img --go
```
on the main README.md page there is UART output shown for the various steps.

it should give a hint as to why step 7 is hanging.

oh, and `fastboot oem uart-on` has basically the same effect as using the uart patched bootloader, but it is hard to run that if you don't have fastboot access :)

**Comment by MaciejMalczyk** [comment]
Thank you for your response. I tried flashing LNX and i was able to use fastboot. Unfortunately i cannot flash bootloader and system image - device hangs while tries to flash image on device. With my knowledge this looks like bad blocks in emmc.

**Comment by tofurky** [comment]
the original n7 i debricked apparently had bad blocks, too, but it seems they were maybe eventually remapped by the controller in the emmc flash chip. they did not prevent reflashing entirely, though.

**Comment by MaciejMalczyk** [comment]
Is there any point of writing with zeros bootloader partition and then trying to flash bootloader? I'm thinking of something simmilar to LNX flash with zeros.

**Comment by tofurky** [comment]
wouldn't hurt if it's already bricked, maybe it could trigger a bad block remap if you try enough times :) writing the LNX partition with zeroes is to trick the bootloader into entering 'fastboot' mode from its current 'Nv3p Server' (nvflash) mode.

**Comment by Mounster-Chef** [comment]
Hello, I don't know if I should make another issue entirely (guessing not), I have an Acer Iconia Tab A510 that has more or less the same problem, except it stops at `sending file: ./bootloader/bootloader-grouper-4.23.img`. I know it isn't quite a Nexus 7 or Ouya, but I thought this guide might do something about my tablet since OpenSurfaceRT also uses it. Is my eMMC fudged?

**Comment by tofurky** [comment]
> Hello, I don't know if I should make another issue entirely (guessing not), I have an Acer Iconia Tab A510 that has more or less the same problem, except it stops at `sending file: ./bootloader/bootloader-grouper-4.23.img`. I know it isn't quite a Nexus 7 or Ouya, but I thought this guide might do something about my tablet since OpenSurfaceRT also uses it. Is my eMMC fudged?

grouper is only for nexus 7 so i don't think it'll run properly. the payload itself should, since the tegra boot rom (burned into the CPU) is the same between devices.

i spent some time looking and while i can find several OTA updates for A510, i can't find a factory image that includes the raw bootloader.

https://www.acer.com/us-en/support/product-support/A510/downloads?suggest=a510;0

there is a file in most of them "bootloader.blob" but it's packed (compressed and/or encrypted):

https://android.googlesource.com/platform/bootable/recovery/+/22d79a5c5eab9c1e86ff2af210bb072689e2d630/bootloader.c#134

```
00000000: 2d53 4947 4e45 442d 4259 2d53 4947 4e42  -SIGNED-BY-SIGNB
00000010: 4c4f 422d bcfb 1600 0001 0000 4d53 4d2d  LOB-........MSM-
00000020: 5241 4449 4f2d 5550 4441 5445 0000 0000  RADIO-UPDATE....
00000030: bcfb 1600 3c00 0000 0100 0000 0000 0000  ....<...........
00000040: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000050: 0000 0000 0000 0000 4542 5400 4c00 0000  ........EBT.L...
...
```
EBT being the bootloader (fastboot) partition from what i recall.

i think the way it works is it's written to USP partition during OTA, then next reboot the bootloader rewrites itself with that file. https://xdaforums.com/t/root-ics-the-hard-way-digging-for-roots.1512787/page-10#post-23337909

edit: i think USP means update/upgrade staging partition

**Comment by earthnoob** [comment]
Hi all. Apologies for trying to make this issue alive again but I've been trying to un-brick my Nexus 7 Wi-Fi (grouper) for the past few days using the method outlined in this repository, however what happened was exactly the same as @MaciejMalczyk encountered, as in, the device became un-responsive after step 7 (even with `--sync` flag added), the terminal stood still, tablet went straight into APX mode after force reboot.

I did nuke the LNX partition as mentioned from one of the comments above before proceeding which successfully made the tablet to go into fastboot mode (Start badge + google logo below) and I was able to click around.

What differed from the OP was that when I tried flashing the stock bootloader file from firmware jwr66y (bootloader-grouper-4.23.img), it gave me this:


`ubuntu@ubuntu:~/Factory_Images/nakasi-jwr66y-factory-201de0b2/nakasi-jwr66y$ fastboot flash bootloader bootloader-grouper-4.23.img --force -v`
`fastboot: verbose: Do flash bootloader bootloader-grouper-4.23.img`
`fastboot: verbose: target didn't report max-download-size`
`Sending 'bootloader' (2100 KB)                     OKAY [  0.261s]`
`Writing 'bootloader'                               FAILED (remote: '(InvalidState)')`
`fastboot: error: Command failed`

There was a little red line on the top left corner of the tablet saying "Signature mismatch". No matter how many times I tried flashing the same image, the line still appeared.

However, I wasn't able to proceed any further with other stock bootloader images from older firmware versions, even trying to flash with the one in the repository also made the terminal hang.

I also used the BCT dump file generated from step 4, as well as following [This guide](https://xdaforums.com/t/guide-recover-from-corrupted-partition-table-no-bricksafe-img.4680047/). Nothing different happened.

I'm getting a bit frustrated as I don't know what I did was wrong, or whether the use of the BCT dump file from step 4 was corrupted or not.
Is there anything else I can do? Do I need to give any files?

**Comment by tofurky** [comment]
no worries, left the issue open because it's unresolved.

it's been a while since i've looked at any of this... but:

regarding the signature mismatch; it seems the one in that particular release is different than the one in the latest release (which matches the .img in the repo):
```
matt@aquos:~/Downloads$ ls -l nakasi*/*/bootloader-grouper-4.23.img
-rw-r----- 1 matt matt 2151068 Aug 21  2013 nakasi-jwr66y-factory-201de0b2/nakasi-jwr66y/bootloader-grouper-4.23.img
-rw-r----- 1 matt matt 2150992 Apr 28  2015 nakasi-lmy47v-factory-5a0bb059/nakasi-lmy47v/bootloader-grouper-4.23.img
matt@aquos:~/Downloads$ md5sum nakasi*/*/bootloader-grouper-4.23.img
5bdb2e87370cdb1a7ea14bb0c3e21390  nakasi-jwr66y-factory-201de0b2/nakasi-jwr66y/bootloader-grouper-4.23.img
df53028033c9eccf4fe5ba7bc198ce24  nakasi-lmy47v-factory-5a0bb059/nakasi-lmy47v/bootloader-grouper-4.23.img
matt@aquos:~/Downloads$ md5sum ~/devel/ouya/tegra30_debrick/bootloader/bootloader-grouper-4.23.img 
df53028033c9eccf4fe5ba7bc198ce24  /home/matt/devel/ouya/tegra30_debrick/bootloader/bootloader-grouper-4.23.img
```

discussed here: https://xdaforums.com/t/q-jwr66y-nakasi-factory-image-bootloader-problem.2417097/

i think what you've done in using this other bootloader image is caused the flash process to error out with a signature mismatch before it even attempted to write anything to emmc. hence the process failing at different stages. as to why the emmc write is hanging, sorry i never came to a conclusive answer but all else being equal it seems like a hardware issue :(

if you used this other .img for both the initial step (6 in particular) AND tried to flash it per step 7, i imagine it *might* progress further without failing the signature check.

there's another debrick method described here: https://github.com/tofurky/tegra30_debrick?tab=readme-ov-file#Other-Methods

worth trying to see if you end up with a different result.

**Comment by tofurky** [comment]
actually, i don't think the older bootloader will work at all. seems to have some extra header on it (hence the larger size):

```
matt@aquos:~/Downloads$ xxd nakasi-jwr66y-factory-201de0b2/nakasi-jwr66y/bootloader-grouper-4.23.img|head
00000000: 4d53 4d2d 5241 4449 4f2d 5550 4441 5445  MSM-RADIO-UPDATE
00000010: 0000 0100 3c00 0000 3c00 0000 0100 0000  ....<...<.......
00000020: 0000 0000 0000 0000 0000 0000 0000 0000  ................
00000030: 0000 0000 0000 0000 0000 0000 4542 5400  ............EBT.
00000040: 4c00 0000 50d2 2000 0100 0000 1000 00ea  L...P. .........
00000050: 18f0 9fe5 18f0 9fe5 18f0 9fe5 18f0 9fe5  ................
00000060: 18f0 9fe5 18f0 9fe5 18f0 9fe5 18f0 9fe5  ................
00000070: 4880 1080 8005 1680 8405 1680 8005 1680  H...............
00000080: 8005 1680 0000 0000 5405 1680 8005 1680  ........T.......
00000090: 0000 0000 dff0 2fe3 e000 9fe5 01db 80e2  ....../.........
matt@aquos:~/Downloads$ xxd nakasi-lmy47v-factory-5a0bb059/nakasi-lmy47v/bootloader-grouper-4.23.img|head
00000000: 1000 00ea 18f0 9fe5 18f0 9fe5 18f0 9fe5  ................
00000010: 18f0 9fe5 18f0 9fe5 18f0 9fe5 18f0 9fe5  ................
00000020: 18f0 9fe5 4880 1080 8005 1680 8405 1680  ....H...........
00000030: 8005 1680 8005 1680 0000 0000 5405 1680  ............T...
00000040: 8005 1680 0000 0000 dff0 2fe3 e000 9fe5  ........../.....
00000050: 01db 80e2 3360 01eb 0b83 01eb d301 0ef1  ....3`..........
00000060: 00f0 6fe3 d101 0ef1 00f0 6fe3 d201 0ef1  ..o.......o.....
00000070: 00f0 6fe3 d701 0ef1 00f0 6fe3 db01 0ef1  ..o.......o.....
00000080: 00f0 6fe3 df01 0ef1 104f 11ee 014a c4e3  ..o......O...J..
00000090: 024b c4e3 0440 c4e3 0140 c4e3 104f 01ee  .K...@...@...O..
```

```
matt@aquos:~/Downloads$ dd if=nakasi-jwr66y-factory-201de0b2/nakasi-jwr66y/bootloader-grouper-4.23.img bs=76 skip=1 of=noheader.img
28302+1 records in
28302+1 records out
2150992 bytes (2.2 MB, 2.1 MiB) copied, 0.0203282 s, 106 MB/s
matt@aquos:~/Downloads$ xxd noheader.img |head
00000000: 1000 00ea 18f0 9fe5 18f0 9fe5 18f0 9fe5  ................
00000010: 18f0 9fe5 18f0 9fe5 18f0 9fe5 18f0 9fe5  ................
00000020: 18f0 9fe5 4880 1080 8005 1680 8405 1680  ....H...........
00000030: 8005 1680 8005 1680 0000 0000 5405 1680  ............T...
00000040: 8005 1680 0000 0000 dff0 2fe3 e000 9fe5  ........../.....
00000050: 01db 80e2 3360 01eb 0b83 01eb d301 0ef1  ....3`..........
00000060: 00f0 6fe3 d101 0ef1 00f0 6fe3 d201 0ef1  ..o.......o.....
00000070: 00f0 6fe3 d701 0ef1 00f0 6fe3 db01 0ef1  ..o.......o.....
00000080: 00f0 6fe3 df01 0ef1 104f 11ee 014a c4e3  ..o......O...J..
00000090: 024b c4e3 0440 c4e3 0140 c4e3 104f 01ee  .K...@...@...O..
matt@aquos:~/Downloads$ md5sum noheader.img nakasi-lmy47v-factory-5a0bb059/nakasi-lmy47v/bootloader-grouper-4.23.img
df53028033c9eccf4fe5ba7bc198ce24  noheader.img
df53028033c9eccf4fe5ba7bc198ce24  nakasi-lmy47v-factory-5a0bb059/nakasi-lmy47v/bootloader-grouper-4.23.img
```

identical when 76 byte header is removed. in other words, just use the .img in the repo. the signature mismatch is a red herring.

**Comment by tofurky** [comment]
> I also used the BCT dump file generated from step 4, as well as following [This guide](https://xdaforums.com/t/guide-recover-from-corrupted-partition-table-no-bricksafe-img.4680047/). Nothing different happened.
> 
> I'm getting a bit frustrated as I don't know what I did was wrong, or whether the use of the BCT dump file from step 4 was corrupted or not. Is there anything else I can do? Do I need to give any files?

regarding the concern the BCT backup is corrupted: on ubuntu/debian, there is a package named `cbootimage` which includes a utility called `bct_dump`. if you install it and run e.g. `bct_dump BCT_READBACK_N7.BIN` (or whatever you named the file) it'll give a bunch of output if it's a valid BCT. you could compare the output from yours to the one in the repo. i don't think that's the reason the reflash is failing, though; i imagine you'd see the same behavior with the known-good copy in git (`bct/nexus_7_grouper_bct.bin`).

**Comment by 4Sitam4** [comment]
@MaciejMalczyk — your symptom (`sent successfully`, then APX after a reboot) is
what I had, and `--sync` did not help me either. I think the explanation is
that **`--download EBT` never writes the BCT, and `--sync` does not write it
either** — I verified by reading the BCT partition back before and after: byte
for byte identical. So the bootloader lands in EBT, but the BCT still has no
valid entry pointing at it, and the bootrom falls back to APX on the next cold
boot.

What actually worked was `--create`, which does sync the BCT. But it fails
first with:

```
bootloader status: BCT is full; no more bootloaders can be added (code: 16)
```

because `bct/nexus_7_grouper_bct.bin` already declares one bootloader
(`BootLoadersUsed = 1` at offset 3920, with an AES hash computed with someone
else's SBK). Blank the slot:

```python
import struct, pathlib
d = bytearray(pathlib.Path('bct/nexus_7_grouper_bct.bin').read_bytes())
struct.pack_into('<I', d, 3920, 0)      # BootLoadersUsed = 0
d[3924:3924+44] = b'\x00' * 44          # blank BootLoader[0]
pathlib.Path('bct-blank.bin').write_bytes(bytes(d))
```

Then `--create` inserts *your* bootloader and the miniloader recomputes the
hash with your device's own AES engine. After that my tablet cold-booted to the
Google logo and stayed there instead of dropping to APX.

Two things you also need, or `--create` will fail on its own: add
`type=sdmmc` to the `[device]` section of `flash.cfg` (otherwise
`specified device is invalid (code: 6)`), and read your real partition table
with `--resume --getpartitiontable` because the shipped `flash.cfg` sizes do
not match a real grouper.

Full write-up including the partition-table/GPT step that comes after:
https://github.com/tofurky/tegra30_debrick/issues/22
